app/src/androidTest/java/com/github/se/eduverse/ui/DashboardScreenTest.kt

```diff
@@ -805,15 +805,140 @@ class DashboardScreenUiTest {
+    composeTestRule.onNodeWithTag("widget_list").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("add_widget_button").assertIsDisplayed()
+  }
+
+  @Test
+  fun testEmptyState_TransitionToWidgetList() {
+    val emptyViewModel = EmptyStateDashboardViewModel()
+
+    composeTestRule.setContent {
+      DashboardScreen(viewModel = emptyViewModel, navigationActions = mockNavigationActions)
+    }
+    composeTestRule.waitForIdle()
+
+    // Verify initial empty state
+    composeTestRule.onNodeWithTag("empty_dashboard_message").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("add_widget_button").assertDoesNotExist()
+
+    // Add widget through view model directly
+    emptyViewModel.addWidget(Widget("test", "Type1", "Test Widget", "Content", "Owner", 0))
+    composeTestRule.waitForIdle()
+
+    // Verify transition to widget list
+    composeTestRule.onNodeWithTag("empty_dashboard_message").assertDoesNotExist()
+    composeTestRule.onNodeWithTag("widget_list").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("add_widget_button").assertIsDisplayed()
+  }
+
+  @Test
+  fun testWidgetList_TransitionToEmptyState() {
+    setupDashboardScreen()
+    composeTestRule.waitForIdle()
+
+    // Verify initial widget list state
+    composeTestRule.onNodeWithTag("widget_list").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("add_widget_button").assertIsDisplayed()
+
+    // Delete all widgets one by one
+    // First, get the initial number of widgets
+    val initialSize = fakeViewModel.widgetList.value.size
+
+    for (i in 0 until initialSize) {
+      // Always click the first delete icon since they'll shift up after each deletion
+      composeTestRule.onAllNodesWithTag("delete_icon")[0].performClick()
+      composeTestRule.waitForIdle()
+    }
+
+    // Verify transition to empty state
+    composeTestRule.onNodeWithTag("empty_dashboard_message").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("empty_state_add_button").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("add_widget_button").assertDoesNotExist()
+  }
+
+  @Test
+  fun testEmptyState_ButtonOpensDialog() {
+    val emptyViewModel = EmptyStateDashboardViewModel()
+
+    composeTestRule.setContent {
+      DashboardScreen(viewModel = emptyViewModel, navigationActions = mockNavigationActions)
+    }
+    composeTestRule.waitForIdle()
+
+    // Click empty state button
+    composeTestRule.onNodeWithTag("empty_state_add_button").performClick()
+    composeTestRule.waitForIdle()
+
+    // Verify dialog content
+    composeTestRule.onNodeWithText("Add Widget").assertIsDisplayed()
+    composeTestRule.onNodeWithTag("add_common_widget_button").assertIsDisplayed()
+
+    // Click cancel
+    composeTestRule.onNodeWithText("Cancel").performClick()
+    composeTestRule.waitForIdle()
+
+    // Verify back to empty state
+    composeTestRule.onNodeWithTag("empty_dashboard_message").assertIsDisplayed()
+  }
+
```

Comment:
The tests are well-implemented and cover key scenarios effectively, including the initial display of the empty state, transitioning to a widget list after adding a widget, and returning to the empty state after deleting all widgets. They also validate user interactions such as opening the add widget dialog and ensuring proper navigation and UI updates. Great attention to detail!